controllers/osartifact_controller.go

```diff
@@ -19,230 +19,285 @@ package controllers
+	cm.Labels[artifactLabel] = artifact.Name
+	if err := controllerutil.SetOwnerReference(artifact, cm, r.Scheme()); err != nil {
 		return ctrl.Result{Requeue: true}, err
 	}
-
-	logger.Info(fmt.Sprintf("Checking deployment %v", osbuild))
-
-	err = r.createRBAC(ctx, osbuild)
-	if err != nil {
+	if err := r.Create(ctx, cm); err != nil && !apierrors.IsAlreadyExists(err) {
 		return ctrl.Result{Requeue: true}, err
 	}
 
-	desiredJob := r.genJob(osbuild)
-	job, err := r.clientSet.BatchV1().Jobs(req.Namespace).Get(ctx, desiredJob.Name, metav1.GetOptions{})
-	if job == nil || apierrors.IsNotFound(err) {
-		logger.Info(fmt.Sprintf("Creating Job %v", job))
-
-		_, err = r.clientSet.BatchV1().Jobs(req.Namespace).Create(ctx, desiredJob, metav1.CreateOptions{})
-		if err != nil {
-			logger.Error(err, "Failed while creating job")
-			return ctrl.Result{}, nil
-		}
-
-		return ctrl.Result{Requeue: true}, nil
+	pvc := r.newArtifactPVC(artifact)
+	if pvc.Labels == nil {
+		pvc.Labels = map[string]string{}
 	}
-	if err != nil {
+	pvc.Labels[artifactLabel] = artifact.Name
+	if err := controllerutil.SetOwnerReference(artifact, pvc, r.Scheme()); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	if err := r.Create(ctx, pvc); err != nil {
 		return ctrl.Result{Requeue: true}, err
 	}
 
-	logger.Info(fmt.Sprintf("Updating state %v", osbuild))
-
-	copy := osbuild.DeepCopy()
-
-	helper, err := patch.NewHelper(&osbuild, r.Client)
-	if err != nil {
-		return ctrl.Result{}, err
+	pod := r.newBuilderPod(pvc.Name, artifact)
+	if pod.Labels == nil {
+		pod.Labels = map[string]string{}
 	}
-	if job.Status.Succeeded > 0 {
-		copy.Status.Phase = "Ready"
-	} else if copy.Status.Phase != "Building" {
-		copy.Status.Phase = "Building"
+	pod.Labels[artifactLabel] = artifact.Name
+	if err := controllerutil.SetOwnerReference(artifact, pod, r.Scheme()); err != nil {
+		return ctrl.Result{Requeue: true}, err
 	}
-	ctx, cancel := context.WithTimeout(context.Background(), 180*time.Second)
-	defer cancel()
-	if err := helper.Patch(ctx, copy); err != nil {
-		return ctrl.Result{}, errors.Wrapf(err, "couldn't patch osbuild %q", copy.Name)
+	if err := r.Create(ctx, pod); err != nil {
+		return ctrl.Result{Requeue: true}, err
 	}
 
-	// for _, c := range append(pod.Status.ContainerStatuses, pod.Status.InitContainerStatuses...) {
-	// 	if c.State.Terminated != nil && c.State.Terminated.ExitCode != 0 {
-	// 		packageBuildCopy.Status.State = "Failed"
-	// 	}
-	// }
+	artifact.Status.Phase = osbuilder.Building
+	if err := r.Status().Update(ctx, artifact); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *OSArtifactReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (r *OSArtifactReconciler) checkBuild(ctx context.Context, artifact *osbuilder.OSArtifact) (ctrl.Result, error) {
+	var pods corev1.PodList
+	if err := r.List(ctx, &pods, &client.ListOptions{
```

Comment:
I'm getting this error:

```
│ manager E0504 13:31:44.393267       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.0/tools/cache/reflector.go:167: Failed to watch *v1.Pod: failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:osar │
│ tifactbuilder-operator-system:osartifactbuilder-operator-controller-manager" cannot list resource "pods" in API group "" at the cluster scope 
```
I'm not sure but I think it comes from here.

Reply:
I don't understand what's going on:

```
~ $ kubectl auth can-i --as system:serviceaccount:osartifactbuilder-operator-system:osartifactbuilder-operator-controller-manager --list
Resources                                       Non-Resource URLs                     Resource Names   Verbs
...
pods.v1                                         []                                    []               [create delete get list watch]
...
~ $ kubectl auth can-i --as system:serviceaccount:osartifactbuilder-operator-system:osartifactbuilder-operator-controller-manager list pods
no
```

Reply:
Realized when I run:

```
~ $ kubectl auth can-i --as system:serviceaccount:osartifactbuilder-operator-system:osartifactbuilder-operator-controller-manager get pods.v1
Warning: the server doesn't have a resource type 'secrets' in group 'v1'

no
```

the same is probably true for secrets in the same file.

Reply:
good catch, I was testing with admin rbac.